src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -2205,28 +2205,31 @@ module GeneralizationHelpers =
 //-------------------------------------------------------------------------
 
 let ComputeInlineFlag (memFlagsOption: SynMemberFlags option) isInline isMutable g attrs m =
-    let hasNoCompilerInliningAttribute() = HasFSharpAttribute g g.attrib_NoCompilerInliningAttribute attrs
-    let enforceNoInlining() = hasNoCompilerInliningAttribute() || HasMethodImplNoInliningAttribute g attrs
-
-    let inlineFlag =
-        let isCtorOrAbstractSlot =
-            match memFlagsOption with
-            | None -> false
-            | Some x -> (x.MemberKind = SynMemberKind.Constructor) || x.IsDispatchSlot || x.IsOverrideOrExplicitImpl
+    let hasNoCompilerInliningAttribute() = HasFSharpAttribute g g.attrib_NoCompilerInliningAttribute attrs  
+    let isCtorOrAbstractSlot() =
+        match memFlagsOption with
+        | None -> false
+        | Some x -> (x.MemberKind = SynMemberKind.Constructor) || x.IsDispatchSlot || x.IsOverrideOrExplicitImpl
 
+    let inlineFlag, reportIncorrectInlineKeywordUsage =
         // Mutable values may never be inlined
         // Constructors may never be inlined
         // Calls to virtual/abstract slots may never be inlined
-        // Values marked with NoCompilerInliningAttribute may never be inlined
-        if isMutable || isCtorOrAbstractSlot || enforceNoInlining() then
-            ValInline.Never 
+        // Values marked with NoCompilerInliningAttribute or [<MethodImpl(MethodImplOptions.NoInlining)>] may never be inlined
+        if isMutable || isCtorOrAbstractSlot() || hasNoCompilerInliningAttribute() then
+            ValInline.Never, errorR
+        elif HasMethodImplNoInliningAttribute g attrs then
+            ValInline.Never, 
+                if g.langVersion.SupportsFeature LanguageFeature.WarningWhenInliningMethodImplNoInlineMarkedFunction
+                then warning
+                else ignore
         elif isInline then
-            ValInline.Always 
+            ValInline.Always, ignore
         else
-            ValInline.Optional
+            ValInline.Optional, ignore
 
     if isInline && (inlineFlag <> ValInline.Always) then 
-        errorR(Error(FSComp.SR.tcThisValueMayNotBeInlined(), m))
+        reportIncorrectInlineKeywordUsage (Error(FSComp.SR.tcThisValueMayNotBeInlined(), m))
```

Comment:
I think at this point `(inlineFlag <> ValInline.Always)` above is no longer needed. Overall this became hard to read now IMO, but no strong feelings, feel free to resolve this comment if you disagree.

Reply:
It is not equivalent, as it processes the HasMethodImplNoInliningAttribute again for a second time.
Also, if something would be e.g. both isCtorOrAbstractSlot AND HasMethodImplNoInliningAttribute  => should be an error, not a warning.

I had something similar to what you have (i.e. repeating the if-then logic in the diagnosticla part), but then reverted it to:
- call the functions at max once
- keep the property of short circuiting ( via || and via if-then) to not read an attribute when it can be avoided.

Reply:
You're right, it is not the same, it's kinda hard to follow. But as I said, I don't feel very strongly about it either way